lib/ipaddr.ml

```diff
@@ -453,87 +453,95 @@ module V4 = struct
   end)
 end
 
-module B128 = struct
-  let int_of_hex_char c =
-    match c with
-    | '0' .. '9' -> Char.code c - 48
-    | 'a' .. 'f' -> Char.code c - 87
-    | 'A' .. 'F' -> Char.code c - 55
-    | _ -> invalid_arg "char is not a valid hex digit"
-
+module S128 : sig
   exception Overflow
 
-  type t = Bytes.t
+  type t
+
+  val zero : t [@@ocaml.warning "-32"]
+  val max_int : t
+  val compare : t -> t -> int
+  val equal : t -> t -> bool
+  val fold_left : ('a -> int -> 'a) -> 'a -> t -> 'a
+  val of_octets_exn : string -> t
+  val to_octets : t -> string
+  val of_int64 : int64 * int64 -> t
+  val to_int64 : t -> int64 * int64
+  val of_int32 : int32 * int32 * int32 * int32 -> t
+  val to_int32 : t -> int32 * int32 * int32 * int32
+  val of_int16 : int * int * int * int * int * int * int * int -> t
+  val to_int16 : t -> int * int * int * int * int * int * int * int
+  val add_exn : t -> t -> t [@@ocaml.warning "-32"]
+  val pred_exn : t -> t [@@ocaml.warning "-32"]
+  val add : t -> t -> t option
+  val logand : t -> t -> t
+  val logor : t -> t -> t
+  val logxor : t -> t -> t
+  val lognot : t -> t
+
+  module Byte : sig
+    val get_lsbits : int -> int -> int [@@ocaml.warning "-32"]
+    val get_msbits : int -> int -> int [@@ocaml.warning "-32"]
+    val set_msbits : int -> int -> int -> int [@@ocaml.warning "-32"]
+    val fold_left : ('a -> bool -> 'a) -> 'a -> int -> 'a
+  end
+
+  val shift_right : t -> int -> t
+  val shift_left : t -> int -> t
+  val write_octets_exn : ?off:int -> t -> bytes -> unit
+  val succ_exn : t -> t
+  val succ : t -> (t, [> `Msg of string ]) result
+  val pred : t -> (t, [> `Msg of string ]) result
+end = struct
+  exception Overflow
 
-  let zero () = Bytes.make 16 '\x00'
-  let max_int () = Bytes.make 16 '\xff'
-  let compare = Bytes.compare
-  let equal = Bytes.equal
+  type t = string
 
-  let fold_left f a b =
-    let a' = ref a in
-    for i = 0 to 15 do
-      let x' = Bytes.get_uint8 b i in
-      a' := f !a' x'
-    done;
-    !a'
+  let mk_zero () = Bytes.make 16 '\x00'
+  let mk_max_int () = Bytes.make 16 '\xff'
+  let zero = Bytes.unsafe_to_string (mk_zero ())
+  let max_int = Bytes.unsafe_to_string (mk_max_int ())
+  let compare = String.compare
+  let equal = String.equal
+  let fold_left f = String.fold_left (fun acc c -> f acc (Char.code c))
```

Comment:
I missed that `String.fold_left` is only since OCaml 4.13.0 :/

Reply:
Reverted `fold_left` to a loop with a comment that `String.fold_left` is available in ocaml>=4.13.0.